### PR TITLE
embeddings,object_detector: add GPU OOM recovering fix

### PR DESCRIPTION
**Pull request by vicentebolea**
fixes: #85

### CODE REVIEW

**Comment by vicentebolea** [comment]
@PaulHax can you check if this change fix your OOM errors while computing embeddings?

**Review by PaulHax — COMMENTED** [review]
I'm getting the error in  `object_detector.py`

```
ERROR:root:Exception raised by task = <Task finished name='Task-71' coro=<TransformsApp._update_images() done, defined at /home/paul/src/cdao/nrtk-explorer/src/nrtk_explorer/app/transforms.py:267> exception=OutOfMemoryError('CUDA out of memory. Tried to allocate 1.22 GiB. GPU ')>
Traceback (most recent call last):
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/trame_server/utils/asynchronous.py", line 19, in handle_task_result
    task.result()
  File "/home/paul/src/cdao/nrtk-explorer/src/nrtk_explorer/app/transforms.py", line 293, in _update_images
    self.compute_predictions_source_images(self.state.source_image_ids)
  File "/home/paul/src/cdao/nrtk-explorer/src/nrtk_explorer/app/transforms.py", line 244, in compute_predictions_source_images
    annotations = self.compute_annotations(ids)
  File "/home/paul/src/cdao/nrtk-explorer/src/nrtk_explorer/app/transforms.py", line 194, in compute_annotations
    predictions = self.detector.eval(image_ids=ids, content=self.context.image_objects)
  File "/home/paul/src/cdao/nrtk-explorer/src/nrtk_explorer/library/object_detector.py", line 80, in eval
    predictions_in_baches = [
  File "/home/paul/src/cdao/nrtk-explorer/src/nrtk_explorer/library/object_detector.py", line 83, in <listcomp>
    self.pipeline(images, batch_size=batch_size),
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/pipelines/object_detection.py", line 104, in __call__
    return super().__call__(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/pipelines/base.py", line 1224, in __call__
    outputs = list(final_iterator)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/pipelines/pt_utils.py", line 124, in __next__
    item = next(self.iterator)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/pipelines/pt_utils.py", line 125, in __next__
    processed = self.infer(item, **self.params)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/pipelines/base.py", line 1150, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/pipelines/object_detection.py", line 117, in _forward
    outputs = self.model(**model_inputs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/models/detr/modeling_detr.py", line 1435, in forward
    outputs = self.model(
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/models/detr/modeling_detr.py", line 1270, in forward
    features, object_queries_list = self.backbone(pixel_values, pixel_mask)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/models/detr/modeling_detr.py", line 411, in forward
    out = self.conv_encoder(pixel_values, pixel_mask)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/transformers/models/detr/modeling_detr.py", line 389, in forward
    features = self.model(pixel_values) if self.config.use_timm_backbone else self.model(pixel_values).feature_maps
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/timm/models/_features.py", line 345, in forward
    return list(self._collect(x).values())
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/timm/models/_features.py", line 299, in _collect
    x = module(x)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1532, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1541, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 460, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/home/paul/src/cdao/.venv/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 456, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 1.22 GiB. GPU 
```

**Comment by vicentebolea** [comment]
I see, the original issue was about the object detector not embedding, these changes are for the embedding extractor, still you can see the error if you increase the batch size and this logic handles that, saying that I will also add the same logic for the object_detector.

**Comment by vicentebolea** [comment]
@PaulHax that was easy to move that logic to the object_detector :smile: , please try this last push

File: src/nrtk_explorer/library/object_detector.py [hunk 1]
```diff
@@ -77,17 +79,38 @@ def eval(
...
+                else:
+                    raise
+
+            finally:
+                # Pytorch needs to freed its allocations outside of the exception context
+                gc.collect()
+                torch.cuda.empty_cache()
+
+        # We should never reach here
+        return None
```
Comment (PaulHax) [inline src/nrtk_explorer/library/object_detector.py:115]: Should we throw?

**Review by PaulHax — APPROVED** [review]
This fixes it!  The back-off is slow though.  With ORID database my comp can only do batch size 8 and it takes a while to get there.  But might be changing processing pipeline alot soon, so maybe not worth investing in right now.

File: src/nrtk_explorer/library/object_detector.py (hunk 1 above)
Comment (vicentebolea) [reply src/nrtk_explorer/library/object_detector.py:115]: we already throw few line above in this execution path.

Comment (vicentebolea) [reply src/nrtk_explorer/library/object_detector.py:115]: maybe this is a c++ ism, maybe we can skip the return none

File: src/nrtk_explorer/library/object_detector.py [hunk 2]
```diff
@@ -77,17 +78,38 @@ def eval(
...
+
+                predictions_by_image_id = {
+                    image_id: predictions
+                    for batch in predictions_in_baches
+                    for image_id, predictions in batch
+                }
+                return predictions_by_image_id
+
+            except RuntimeError as e:
+                if "out of memory" in str(e) and adjusted_batch_size > 2:
```
Comment (PaulHax) [inline src/nrtk_explorer/library/object_detector.py:100]:
```suggestion
                if "out of memory" in str(e) and adjusted_batch_size >= 2:
```

File: src/nrtk_explorer/library/object_detector.py [hunk 3]
```diff
@@ -77,17 +78,38 @@ def eval(
...
+                    previous_batch_size = adjusted_batch_size
+                    adjusted_batch_size = max(2, adjusted_batch_size // 2)
```
Comment (PaulHax) [inline src/nrtk_explorer/library/object_detector.py:102]:
```suggestion
                    adjusted_batch_size =  adjusted_batch_size // 2
```
We want to allow batch size of 1.

**Comment by vicentebolea** [comment]
@PaulHax It appears that the object_detection failure with batch size 1 was about something else. Find the following changes in the latest push:

- I added a Quasar input for selection object detection batch_size in the annotations section.
- I allowed to go to batch_size=1 before rethrowing the OOM exception. I cannot see the batch_size=1 error anymore.

File: src/nrtk_explorer/library/object_detector.py (hunk 1 above)
Comment (vicentebolea) [reply src/nrtk_explorer/library/object_detector.py:115]: Apparently mypy does not like not having a return statement even if it is legal in python. `src/nrtk_explorer/library/object_detector.py:59: error: Missing return statement  [return]`

**Review by PaulHax — APPROVED** [review]
Verks!

**Comment by vicentebolea** [comment]
> Verks!

Yay!

**Pull request merged**